security/rbac/engine/engine.go

```diff
@@ -0,0 +1,197 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package engine
+
+import (
+	pb "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v2"
+	cel "github.com/google/cel-go/cel"
+	interpreter "github.com/google/cel-go/interpreter"
+	expr "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+)
+
+// AuthorizationArgs is the input of the CEL-based authorization engine.
+type AuthorizationArgs struct {
+	md       metadata.MD
+	peerInfo *peer.Peer
+}
+
+// Converts AuthorizationArgs into the activation for CEL.
+func (args *AuthorizationArgs) toActivation() *interpreter.Activation {
+	// TODO(@ezou): implement the conversion logic.
+	return nil
+}
+
+// Decision is the enum type that represents different authorization
+// decisions a CEL-based authorization engine can return.
+type Decision int32
+
+const (
+	// DecisionAllow indicates allowing the RPC to go through.
+	DecisionAllow Decision = iota
+	// DecisionDeny indicates denying the RPC from going through.
+	DecisionDeny
+	// DecisionUnknown indicates that there is insufficient information to
+	// determine whether or not an RPC call is authorized.
+	DecisionUnknown
+)
+
+// String returns the string representation of a Decision object.
+func (d Decision) String() string {
+	return [...]string{"DecisionAllow", "DecisionDeny", "DecisionUnknown"}[d]
+}
+
+// AuthorizationDecision is the output of CEL-based authorization engines.
+// If decision is allow or deny, policyNames will either contain the names of
+// all the policies matched in the engine that permitted the action, or be
+// empty as the decision was made after all conditions evaluated to false.
+// If decision is unknown, policyNames will contain the list of policies that
+// evaluated to unknown.
+type AuthorizationDecision struct {
+	decision    Decision
+	policyNames []string
+}
+
+// Converts an expression to a parsed expression, with SourceInfo nil.
+func exprToParsedExpr(condition *expr.Expr) *expr.ParsedExpr {
+	return &expr.ParsedExpr{Expr: condition}
+}
+
+// Converts an expression to a CEL program.
+func exprToProgram(condition *expr.Expr) *cel.Program {
+	// TODO(@ezou): implement the conversion from expr to CEL program.
+	return nil
+}
+
+// Evaluates a CEL expression. Returns true if CEL evaluation returns true. Returns false if
+// CEL evaluation returns false. Returns an error if CEL evaluation returns unknown or an
+// unsuccessful evaluation.
+func evaluateProgram(program *cel.Program, activation *interpreter.Activation) (bool, error) {
```

Comment:
Why do we want to use an activation here? If we use interface{}, we can pass wither  a map or an activation and the CEL library will choose the right way for us. If we only accept Activation, how would a user do if they want to pass a map?

Reply:
Also cel.Program already defined a similar function:
```
type Program interface {
	Eval(vars interface{}) (ref.Val, *EvalDetails, error)
}
```
Adding one more layer of this evaluation logic seems unnecessary here(I could be wrong). Can we just reuse this interface here? This might also help when writing unit tests.

Reply:
![Screenshot 2020-07-24 at 9 33 30 AM](https://user-images.githubusercontent.com/27510384/88402461-c312df00-cd90-11ea-97e4-4b8cf8727b12.png)

Reply:
Hmm I just changed it from my understanding of @jiangtaoli2016 's comment, maybe we can keep both..?

Reply:
According to Tristan, Activation is more efficient. We should use Activation -- also consistent with other languages.